File: patterns/typescript/slidingWindow.ts
class SlidingWindow {
    // Brute Force Approach - O(n * k)
    findMaxAverageBruteForce(nums: number[], k: number): number {
        let maxAvg = -Infinity;

        for (let i = 0; i <= nums.length - k; i++) {
            let sum = 0;
            for (let j = i; j < i + k; j++) {
                sum += nums[j];
            }
            maxAvg = Math.max(maxAvg, sum / k);
        }
        return maxAvg;
    }

    // Sliding Window Approach - O(n)
    findMaxAverageSlidingWindow(nums: number[], k: number): number {
        let sum = nums.slice(0, k).reduce((a, b) => a + b, 0);
        let maxSum = sum;

        for (let i = k; i < nums.length; i++) {
            sum += nums[i] - nums[i - k];
            maxSum = Math.max(maxSum, sum);
        }

        return maxSum / k;
    }

    // Sliding Window for Longest Substring Without Repeating Characters
    lengthOfLongestSubstringSlidingWindow(s: string): number {
        let seen = new Set<string>();
        let maxLength = 0, left = 0;

        for (let right = 0; right < s.length; right++) {
            while (seen.has(s[right])) {
                seen.delete(s[left]);
                left++;
            }
            seen.add(s[right]);
            maxLength = Math.max(maxLength, right - left + 1);
        }
        return maxLength;
    }

    // Sliding Window using Frequency Array
    lengthOfLongestSubstringSlidingWindowFrequencyArray(s: string): number {
        let freq = new Array(128).fill(0);
        let maxLength = 0, left = 0;

        for (let right = 0; right < s.length; right++) {
            freq[s.charCodeAt(right)]++;

            while (freq[s.charCodeAt(right)] > 1) {
                freq[s.charCodeAt(left)]--;
                left++;
            }

            maxLength = Math.max(maxLength, right - left + 1);
        }
        return maxLength;
    }
}

File: patterns/go/sliding_window.go
package main

import (
	"math"
)

// Brute Force Approach - O(n * k)
func findMaxAverageBruteForce(nums []int, k int) float64 {
	n := len(nums)
	maxAvg := math.Inf(-1)

	for i := 0; i <= n-k; i++ {
		sum := 0
		for j := i; j < i+k; j++ {
			sum += nums[j]
		}
		maxAvg = math.Max(maxAvg, float64(sum)/float64(k))
	}
	return maxAvg
}

// Sliding Window Approach - O(n)
func findMaxAverageSlidingWindow(nums []int, k int) float64 {
	sum := 0
	for i := 0; i < k; i++ {
		sum += nums[i]
	}

	maxSum := sum

	for i := k; i < len(nums); i++ {
		sum += nums[i] - nums[i-k]
		if sum > maxSum {
			maxSum = sum
		}
	}

	return float64(maxSum) / float64(k)
}

// Sliding Window for Longest Substring Without Repeating Characters
func lengthOfLongestSubstringSlidingWindow(s string) int {
	seen := make(map[byte]bool)
	maxLength, left := 0, 0

	for right := 0; right < len(s); right++ {
		for seen[s[right]] {
			delete(seen, s[left])
			left++
		}
		seen[s[right]] = true
		maxLength = max(maxLength, right-left+1)
	}
	return maxLength
}

// Sliding Window using Frequency Array
func lengthOfLongestSubstringSlidingWindowFrequencyArray(s string) int {
	freq := make([]int, 128)
	maxLength, left := 0, 0

	for right := 0; right < len(s); right++ {
		freq[s[right]]++

		for freq[s[right]] > 1 {
			freq[s[left]]--
			left++
		}

		maxLength = max(maxLength, right-left+1)
	}
	return maxLength
}

// Helper function to get max of two numbers
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
